### ISSUE TITLE
Movie loop's lag

### ISSUE THREAD

**Issue by codedealer** [renpy/renpy#770]
**tl;dr** Cache/buffer videos pls?

That problem was skimmed on the forum a few years ago, I think. Video playback performance is poor when it comes to loops. What it does is basically skips a few frames of video hence breaking seamless transition.
As far as I can tell this problem is OS agnostic and has been around for a while. Can't RenPy buffer a video file? The code explicitly states that apart from audio files buffering of video is prohibited which seems to be the shatter point.

I understand this may seem like insignificant feature for a visual novel engine, but it's 2015. We have rich multimedia support on so many devices. Plus looping of video files is already baked in the engine but the described flaw kind of cancels it. We have ATL for simple stuff but no gif, apng (you name it) support for rich animation so the video is the only solution. 

The most basic use case of it is looping video background in the main menu — a question that pops up on the forum every now and then.

**Comment by renpytom** [comment]
Honestly, I barely understand the movie playback code - it's taken from ffplayer. If you want to submit a fix, I would be happy to accept it, but I can't even reproduce the problem here.

**Comment by codedealer** [comment]
Too bad, I hardly understand that either but I can at least promise you I will look more into that.

**Comment by vollschauer** [comment]
Oh yes I already have those loop delays (lags) on my system (linux). How can we help? Where do I have to dig?

**Comment by vollschauer** [comment]
I have captured this issue to make sure we are talking about the same.
https://www.youtube.com/watch?v=rw26Tv8RIAc

**Comment by codedealer** [comment]
@vollschauer I can confirm this is exactly I what was talking about. From what I've determined it has something to do with the flawed ffmpeg code mentioned before and not renpy as it is. 
Funny fact I was testing movie textures in Unity the other day. This is the only way to play video in Unity so far. And guess what upon looping a movie texture video playback skips 6-7 frames similar to those in renpy. 

I **would** say that is because of theora ogv codec specifics since I was testing ogv videos and they are supported by both Unity and renpy, but I can see you tested webm as well.
So my assumption is that Unity and renpy share the same ffplayer code (can't be confirmed though b/c Unity isn't open source) and that code deals terribly with looping videos. I failed to discover what the problem was exactly.

**Comment by vollschauer** [comment]
@renpytom can you confirm this issue or do I/we make something wrong with our movie files (bitrate,codec whatever)?

**Comment by vollschauer** [comment]
@renpytom do you have any working loop movie sample? This is the last TODO on my list to
finish my project/work.... 

@codedealer 
if you check your movies with the tool "mediainfo"...do you have any difference between the video and audio duration of your clip?

**Comment by codedealer** [comment]
@vollschauer hm, good guess. Video track is 10 ms longer than audio. This could possibly desync video file but... 10 ms? I checked my files with after effects and didn't even notice it.

**Comment by vollschauer** [comment]
Spend a lot of hours into it today and there is no way I could make this work with ogv or webm. So I went back to my old mpg to find out why they are not playing at all (they stuttering like hell.. ) It seems RenPy doesn't like it if the mpeg has B-Frame's enabled. I still need to find a way to get rid of the 5ms time difference between video/audio. But I don't have this loop lag with mpeg-1 or mpeg-2.

**Comment by renpytom** [comment]
This has been fixed by the new-audio branch, which you can get from http://nightly.renpy.org, and is going to be released in 6.99.9 shortly.

**Comment by vollschauer** [comment]
Thank you! Was looking forward to get this one solved. Is there any big change how to play movies inside displayables? I read the doc's from the latest nightly but it wouldn't play for example ogv files. 
I also checked in the tutorial the movie section, this one doesn't play either.

**Comment by vollschauer** [comment]
Anyway got everything working now :) I'm right we don't need a silent audio track on movies?
